pkg/utils/kubernetes/eventhandler.go

```diff
@@ -0,0 +1,328 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// ControllerPredicateFactory provides a method for creating new Predicates for a controller.
+type ControllerPredicateFactory interface {
+	// NewControllerPredicate creates and returns a new Predicate with the given controller.
+	NewControllerPredicate(client.Object) predicate.Predicate
+}
+
+// ControllerPredicateFactoryFunc is a function that implements ControllerPredicateFactory.
+type ControllerPredicateFactoryFunc func(client.Object, client.Object, bool) bool
+
+// NewControllerPredicate creates and returns a new Predicate with the given controller.
+func (f ControllerPredicateFactoryFunc) NewControllerPredicate(controller client.Object) predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return f(e.Object, controller, false)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return f(e.ObjectNew, controller, false)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return f(e.Object, controller, true)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+}
+
+// Enqueuer provides a method for enqueuing an object for processing.
+type Enqueuer interface {
+	// Enqueue enqueues the object for processing.
+	Enqueue(client.Object)
+}
+
+// EnqueuerFunc is a function that implements Enqueuer.
+type EnqueuerFunc func(client.Object)
+
+// Enqueue enqueues the object for processing.
+func (f EnqueuerFunc) Enqueue(obj client.Object) {
+	f(obj)
+}
+
+// ControlledResourceEventHandler is an implementation of cache.ResourceEventHandler that enqueues controllers upon
+// controlled resource events, if a predicate check is successful.
+type ControlledResourceEventHandler struct {
+	// ControllerTypes is a list of controller types. If multiple controller types are specified,
+	// there is a chain of controllers. Only the top-most controller (with the last listed type) is eventually enqueued.
+	ControllerTypes []ControllerType
+	// Ctx is the context used to get controller objects.
+	Ctx context.Context
+	// Client is the client used to get controller objects.
+	Client client.Client
```

Comment:
Wouldn't `Client` use a cache as well? What if the object we are interesting in is not already in the cache (or not updated), wouldn't then a cache-based reader miss the object or fetch an old state?

Reply:
> Wouldn't client.Client use a cache as well (if got via Client()) ?

Hopefully we will get there soon. But ATM, if the `CachedRuntimeClients` feature gate is disabled, `Client()` will be uncached.

> What if the object we are interested in is not already in the cache (or out of date)

The changed object itself will always be in the cache already, as it's based on the same watch as the event handler (at least that's true for your controller, which only relies on c-r informers).

The owning object will almost certainly also be in the cache already. Because if it wasn't in the cache already, no controller would have acted on it yet and no ownees would've been created to trigger any events.

Reply:
OK, I trust you on that, I will change it as suggested.